### PR TITLE
Op#3648 reduce memory consumption in track parser

### PR DESCRIPTION
**Pull request by KueblerJelle**
OP#3648

### CODE REVIEW

File: OTAnalytics/plugin_parser/otvision_parser.py [hunk 1]
```diff
@@ -266,13 +304,25 @@ def __parse_otdet_version(self, content: dict) -> Version:
...
+
+    # TODO Review: fixing metadata alone can be reused in streaming_parser,
+    # TODO hence I made it public. This might also apply to fixing
+    # TODO single detection data, when they are parsed lazy (but currently not used).
+    def fix_metadata(self, metadata: dict) -> dict:
+        """
+        Fix formate changes from older ottrk metadata
+        format versions to the current version.
+        """
         current_version = Version.from_str(metadata[ottrk_format.OTTRK_VERSION])
```
Comment (KueblerJelle) [inline OTAnalytics/plugin_parser/otvision_parser.py:317]: @briemla Review: fixing metadata alone can be reused in streaming_parser, hence I made it public. 
This might also apply to fixing single detection data, when they are parsed lazy (but currently not used).

File: OTAnalytics/plugin_parser/otvision_parser.py [hunk 2]
```diff
@@ -224,11 +224,49 @@ def __fix_tracking_run_ids(self, metadata: dict, current_version: Version) -> di
...
 
 
+class Ottrk_Version_1_0_To_1_2(MetadataFixer):
+    def __init__(self) -> None:
+        super().__init__(VERSION_1_0, VERSION_1_2)
+
+    def fix(self, metadata: dict, current_version: Version) -> dict:
+        return self.__fix_recorded_start_date(metadata, current_version)
+
+    def __fix_recorded_start_date(
```
Comment (KueblerJelle) [inline OTAnalytics/plugin_parser/otvision_parser.py:234]: @briemla Review: Not sure in which version this changed

File: OTAnalytics/plugin_parser/otvision_parser.py [hunk 3]
```diff
@@ -341,6 +391,65 @@ def __str__(self) -> str:
 )
 
 
+# TODO Review: these methods for creating PythonDetections and PythonTracks are static
+# TODO and could live outside a class for reusability in streaming_parser
+def parse_python_detection(
```
Comment (KueblerJelle) [inline OTAnalytics/plugin_parser/otvision_parser.py:389]: @briemla  Review: these methods for creating PythonDetections and PythonTracks are static and could live outside a class for reusability in streaming_parser

File: OTAnalytics/plugin_parser/otvision_parser.py [hunk 4]
```diff
@@ -922,7 +1003,8 @@ def parse(
         return track_ids, [video] * len(track_ids)
 
     def __parse_recorded_start_date(self, metadata: dict) -> datetime:
-        start_date = metadata[ottrk_format.RECORDED_START_DATE]
+        print(metadata)
```
Comment (briemla) [inline OTAnalytics/plugin_parser/otvision_parser.py:1006]: In case we need the print statement, please use `logger().info(...)` instead.

File: OTAnalytics/plugin_parser/track_export.py [hunk 5]
```diff
@@ -22,22 +36,33 @@ def __init__(
...
-        write_json(tracks_metadata, tracks_metadata_path)
-        videos_metadata_path = specification.save_path.with_suffix(
-            ".videos_metadata.json"
-        )
-        videos_metadata = self._videos_metadata.to_dict()
-        write_json(videos_metadata, videos_metadata_path)
+        path = specification.save_path
+        output_path = path.with_suffix(".tracks.csv")
+        write_mode: Literal["w", "a"] = "a" if append else "w"
+        dataframe.to_csv(output_path, index=False, header=(not append), mode=write_mode)
```
Comment (briemla) [inline OTAnalytics/plugin_parser/track_export.py:55]: Parentheses are not necessary.

File: OTAnalytics/plugin_parser/streaming_parser.py [hunk 6]
```diff
@@ -0,0 +1,349 @@
...
+RawVideoMetadata = dict
+RawFileData = tuple[RawDetectionData, RawVideoMetadata, TrackIdGenerator]
+
+
+def parse_json_bz2_events(path: Path) -> Iterable[tuple[str, str, str]]:
+    """
+    Provide lazy data stream reading the bzip2 compressed file
+    at the given path and interpreting it as json objects.
+    """
+    stream = bz2.BZ2File(path)
```
Comment (briemla) [inline OTAnalytics/plugin_parser/streaming_parser.py:48]: Do we need to use the with context manager to ensure the file handle is properly closed at the end of reading?

File: OTAnalytics/plugin_parser/streaming_parser.py [hunk 7]
```diff
@@ -0,0 +1,349 @@
...
+            det = parse_python_detection(
+                metadata_video, id_generator, det_dict, input_file
+            )
+
+            # Group detections by track id
+            if not self._tracks_dict.get(det.track_id):
+                self._tracks_dict[det.track_id] = []
+            self._tracks_dict[det.track_id].append(det)
+
+            # the finished flag indicates tha last detection of a track
```
Comment (briemla) [inline OTAnalytics/plugin_parser/streaming_parser.py:153]: typo in `tha`

File: OTAnalytics/plugin_parser/streaming_parser.py [hunk 8]
```diff
@@ -0,0 +1,349 @@
...
+            tracks_metadata.update_detection_classes(
+                new_detection_metadata.detection_classes
+            )
+
+        for videos_metadata in self._registered_videos_metadata:
+            videos_metadata.update(new_video_metadata)
+
+    def parse(self, files: set[Path]) -> Iterator[TrackDataset]:
+        iterator = iter(self._parse_tracks(files))
+        while True:
```
Comment (briemla) [inline OTAnalytics/plugin_parser/streaming_parser.py:292]: When does the loop end?

File: OTAnalytics/plugin_parser/streaming_parser.py [hunk 9]
```diff
@@ -0,0 +1,349 @@
...
+
+def default_track_dataset_factory(tracks: list[Track]) -> TrackDataset:
+    return PandasTrackDataset.from_list(
+        tracks=tracks,
+        track_geometry_factory=PygeosTrackGeometryDataset.from_track_dataset,
+        calculator=PandasByMaxConfidence(),
+    )
+
+
+class StreamOttrkParser(StreamTrackParser):
```
Comment (briemla) [inline OTAnalytics/plugin_parser/streaming_parser.py:216]: This will be the future main parser for track data. Please add a unit test using our simple example in tests/data. The example has 60 frames. Please test the edge cases reading the first chunk of frames and the last chunk of frames.

File: OTAnalytics/plugin_prototypes/eventlist_exporter/eventlist_exporter.py [hunk 10]
```diff
@@ -136,13 +192,49 @@ def get_name(self) -> str:
...
+        events: Iterable[Event],
+        sections: Iterable[Section],
+        export_specification: EventExportSpecification,
     ) -> None:
-        self._event_list_parser.serialize(events, sections, file)
+        file = export_specification.file
+        export_mode = export_specification.export_mode
+
+        if export_mode == INITIAL_MERGE:
+            print(
```
Comment (briemla) [inline OTAnalytics/plugin_prototypes/eventlist_exporter/eventlist_exporter.py:222]: Please use `logger().info(...)` or `.warning`.

File: tests/utils/assertions.py [hunk 11]
```diff
@@ -63,8 +63,35 @@ def assert_track_dataset_has_tracks(dataset: TrackDataset, tracks: list[Track])
...
-        actual_lines = sorted(actual.readlines())
+        act_lines = sorted(actual.readlines())
         with open(expected_counts_file, mode="r") as expected:
-            expected_lines = sorted(expected.readlines())
-            assert actual_lines == expected_lines
+            exp_lines = sorted(expected.readlines())
+
+            assert set(act_lines) == set(exp_lines), (
+                f"Sets are not equal [act len: {len(act_lines)}, "
+                "exp len {len(expected_lines)}]:\n"
```
Comment (briemla) [inline tests/utils/assertions.py:74]: `f` missing for f-string

File: tests/regression_otanalytics.py [hunk 12]
```diff
@@ -53,11 +59,17 @@ def otflow_parser() -> FlowParser:
 class TestRegressionCompleteApplication:
 
     @pytest.mark.skip
+    @pytest.mark.parametrize(
+        "cli_mode",
+        [CliMode.STREAM, CliMode.BULK],
```
Comment (briemla) [inline tests/regression_otanalytics.py:64]: Test data should only be recreated once as it will result in the same files

File: tests/regression_otanalytics.py [hunk 13]
```diff
@@ -71,14 +83,22 @@ def test_15_min_recreate_test_data(
...
                 otflow_parser=otflow_parser,
                 event_formats=("csv", "otevents"),
+                cli_mode=cli_mode,
+                cli_chunk_size=cli_chunk_size,
             )
 
     @pytest.mark.skip
+    @pytest.mark.parametrize(
+        "cli_mode",
+        [CliMode.STREAM, CliMode.BULK],
```
Comment (briemla) [inline tests/regression_otanalytics.py:93]: Test data should only be recreated once as it will result in the same files

File: tests/regression_otanalytics.py [hunk 14]
```diff
@@ -93,14 +113,22 @@ def test_2_h_single_recreate_test_data(
...
```
Comment (briemla) [inline tests/regression_otanalytics.py:123]: Test data should only be recreated once as it will result in the same files

File: tests/regression_otanalytics.py [hunk 15]
```diff
@@ -115,14 +143,22 @@ def test_2_h_recreate_test_data(
...
```
Comment (briemla) [inline tests/regression_otanalytics.py:153]: Test data should only be recreated once as it will result in the same files

File: tests/OTAnalytics/plugin_ui/test_cli.py [hunk 16]
```diff
@@ -373,6 +407,14 @@ def cli_dependencies(self) -> dict[str, Any]:
                     DEFAULT_TRACK_LENGTH_LIMIT,
                 ),
             ),
+            # self.TRACK_PARSER: StreamOttrkParser(
```
Comment (briemla) [inline tests/OTAnalytics/plugin_ui/test_cli.py:410]: Please remove commented code

File: tests/OTAnalytics/plugin_ui/test_cli.py [hunk 17]
```diff
@@ -529,8 +803,15 @@ def test_start_with_no_video_in_folder(
             save_name=str(save_name),
             save_suffix=save_suffix,
             count_intervals=[count_interval],
+            cli_mode=mode,
         )
-        cli = OTAnalyticsCli(run_config, **cli_dependencies)
+
+        cli: OTAnalyticsCli
+        if mode == CliMode.STREAM:
```
Comment (briemla) [inline tests/OTAnalytics/plugin_ui/test_cli.py:810]: This if statement is duplicated in multiple tests. Please consider extracting it into a method or function

File: tests/OTAnalytics/plugin_ui/test_cli.py [hunk 18]
```diff
@@ -291,12 +307,30 @@ def mock_cli_dependencies(self) -> dict[str, Any]:
...
+            self.TRACK_PARSER: Mock(spec=StreamTrackParser),
+            **mock_cli_dependencies,
+        }
+
+    @pytest.fixture
+    def mock_cli_bulk_dependencies(
+        self, mock_cli_dependencies: dict[str, Any]
+    ) -> dict[str, Any]:
+        return {
+            self.PROGRESSBAR: Mock(spec=NoProgressbarBuilder),
```
Comment (briemla) [inline tests/OTAnalytics/plugin_ui/test_cli.py:328]: Why is `self.PROGRESSBAR` only set for bulk_dependencies?

File: tests/OTAnalytics/plugin_ui/test_cli.py [hunk 19]
```diff
@@ -391,57 +433,283 @@ def cli_dependencies(self) -> dict[str, Any]:
...
+        return {
+            self.TRACK_PARSER: OttrkParser(
+                PythonDetectionParser(
+                    ByMaxConfidence(),
+                    tracks_metadata._track_repository,
+                    PygeosTrackGeometryDataset.from_track_dataset,
+                    DEFAULT_TRACK_LENGTH_LIMIT,
+                ),
+            ),
+            self.PROGRESSBAR: NoProgressbarBuilder(),
```
Comment (briemla) [inline tests/OTAnalytics/plugin_ui/test_cli.py:556]: Why is `self.PROGRESSBAR` only set for bulk_dependencies?

File: tests/OTAnalytics/plugin_ui/test_cli.py [hunk 20]
```diff
@@ -276,7 +292,7 @@ class TestOTAnalyticsCli:
     @pytest.fixture
     def mock_cli_dependencies(self) -> dict[str, Any]:
         return {
-            self.TRACK_PARSER: Mock(spec=TrackParser),
+            # self.TRACK_PARSER: Mock(spec=TrackParser),
```
Comment (briemla) [inline tests/OTAnalytics/plugin_ui/test_cli.py:295]: Please remove commented code.

File: tests/OTAnalytics/plugin_ui/test_cli.py [hunk 21]
```diff
@@ -291,12 +307,30 @@ def mock_cli_dependencies(self) -> dict[str, Any]:
             self.CLEAR_ALL_TRACKS: Mock(spec=ClearAllTracks),
             self.TRACKS_METADATA: Mock(spec=TracksMetadata),
             self.VIDEOS_METADATA: Mock(spec=VideosMetadata),
-            self.PROGRESSBAR: Mock(spec=NoProgressbarBuilder),
             self.EXPORT_ROAD_USER_ASSIGNMENT: Mock(spec=ExportRoadUserAssignments),
+            # self.PROGRESSBAR: Mock(spec=ProgressbarBuilder),
```
Comment (briemla) [inline tests/OTAnalytics/plugin_ui/test_cli.py:311]: Please remove commented code.

File: OTAnalytics/plugin_ui/customtkinter_gui/dummy_viewmodel.py [hunk 22]
```diff
@@ -6,6 +6,8 @@
 from tkinter.filedialog import askopenfilename, askopenfilenames
 from typing import Any, Iterable, Optional
 
+from application.export_formats.export_mode import OVERWRITE
```
Comment (briemla) [inline OTAnalytics/plugin_ui/customtkinter_gui/dummy_viewmodel.py:9]: Please add `OTAnalytics` before `application` to be consistent with the other imports

File: OTAnalytics/plugin_ui/cli.py [hunk 23]
```diff
@@ -281,14 +294,169 @@ def _do_export_counts(self, save_path: Path) -> None:
...
+        self._track_parser.register_videos_metadata(self._videos_metadata)
+
+        return self._track_parser.parse(track_files)
+
+    def _run_analysis(self, ottrk_files: set[Path]) -> None:
+        """Run analysis."""
+        sections = self._run_config.sections
+        is_first = True
+
+        track_stream = peekable(self._parse_track_stream(ottrk_files))
```
Comment (briemla) [inline OTAnalytics/plugin_ui/cli.py:435]: Nice one. I didn't know about that one :-)

**Review by briemla — CHANGES_REQUESTED** [review]
Looks good. As this is a larger PR, I have few questions and comments.

File: OTAnalytics/plugin_parser/otvision_parser.py (hunk 1 above)
Comment (KueblerJelle) [reply OTAnalytics/plugin_parser/otvision_parser.py]: Fixed in commit 4aaf9a5, metadatafixer can now define whether they are based on otdet or ottrk format version

File: OTAnalytics/application/datastore.py [hunk 24]
```diff
@@ -317,7 +321,11 @@ def export_event_list_file(
         event_list_exporter.export(
             events=self._event_repository.get_all(),
             sections=self._section_repository.get_all(),
-            file=file,
+            export_specification=EventExportSpecification(
+                file=file,
+                format=event_list_exporter.get_extension(),
```
[the remark this answers is not in the document]
Comment (KueblerJelle) [reply OTAnalytics/application/datastore.py]: True, 'format' Attribute is now removed from EventExportSpecification (359d9a3)

File: OTAnalytics/plugin_parser/export.py [hunk 25]
```diff
@@ -37,22 +39,40 @@
...
-            return
-        dataframe = self._add_detailed_date_time_columns(dataframe)
-        dataframe = self._set_column_order(dataframe)
-        dataframe = dataframe.sort_values(
-            by=[LEVEL_START_TIME, LEVEL_END_TIME, LEVEL_CLASSIFICATION]
-        )
-        dataframe.to_csv(self.__create_path(), index=False)
-        logger().info(f"Counts saved at {self._output_file}")
+
+        for tag, value in counts.to_dict().items():
```
[the remark this answers is not in the document]
Comment (KueblerJelle) [reply OTAnalytics/plugin_parser/export.py:58]: Added test for export modes INITIAL_MERGE and MERGE export modes,
Adapted existing test to test FLUSH additionally to OVERWRITE 
821ac6d

File: OTAnalytics/plugin_parser/streaming_parser.py (hunk 8 above)
Comment (KueblerJelle) [reply OTAnalytics/plugin_parser/streaming_parser.py:292]: once the iterator providing the parsed tracks from the files runs out of tracks, the resulting chunk will be an empty list and the break condition inside the loop will trigger

Comment (briemla) [reply OTAnalytics/plugin_parser/streaming_parser.py:292]: Should we use an empty result here or raise a StopIteration?

Comment (KueblerJelle) [reply OTAnalytics/plugin_parser/streaming_parser.py:292]: Using break and return should both have the same behavior: the generator function ends and a StopIteration is raised. 
But using a return statment, makes the intent / meaning clearer. Changed in 124eeed

File: tests/OTAnalytics/plugin_ui/test_cli.py (hunk 18 above)
Comment (KueblerJelle) [reply tests/OTAnalytics/plugin_ui/test_cli.py:340]: For OTAnalyticsStreamCli, the progressbar shows progress in terms of processed files, as the number of tracks is not known from the lazy stream.
Files are being processed in the StreamOttrkParser implementation, hence the progressbar is a constructor argument of this object and not cli (since the parser itself is a constructor argument of cli)

File: tests/OTAnalytics/plugin_ui/test_cli.py (hunk 19 above)
Comment (KueblerJelle) [reply tests/OTAnalytics/plugin_ui/test_cli.py:560]: For OTAnalyticsStreamCli, the progressbar shows progress in terms of processed files, as the number of tracks is not known from the lazy stream.
Files are being processed in the StreamOttrkParser implementation, hence the progressbar is a constructor argument of this object and not cli (since the parser itself is a constructor argument of cli)

**Review by briemla — DISMISSED** [review]

**Review by briemla — APPROVED** [review]
